python/ml4ir/base/model/optimizer.py

```diff
@@ -37,38 +24,83 @@ def get_optimizer(
     References:
         https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer
         https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/ExponentialDecay
+        https://arxiv.org/pdf/1506.01186.pdf
 
     FIXME:
         Define all arguments overriding tensorflow defaults in a separate file
         for visibility with ml4ir users
     """
-    # Define an exponential learning rate decay schedule
-    learning_rate_schedule = ExponentialDecay(
-        learning_rate,
-        decay_steps=learning_rate_decay_steps,
-        decay_rate=learning_rate_decay,
-        staircase=True,
-    )
+    model_config = file_io.read_yaml(model_config_file)
 
-    if optimizer_key == OptimizerKey.ADAM:
-        return tf_optimizers.Adam(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    elif optimizer_key == OptimizerKey.NADAM:
-        return tf_optimizers.Nadam(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    elif optimizer_key == OptimizerKey.ADAGRAD:
-        return tf_optimizers.Adagrad(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    elif optimizer_key == OptimizerKey.SGD:
-        return tf_optimizers.SGD(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    elif optimizer_key == OptimizerKey.RMS_PROP:
-        return tf_optimizers.RMSprop(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
+    if 'lr_schedule' not in model_config:
+        learning_rate_schedule = ExponentialDecay(
+            initial_learning_rate=0.01,
+            decay_steps=10000000,
+            decay_rate=1.0,
+            staircase=True,
         )
+
+    else:
```

Comment:
@mohazahran your PR does great in terms of functionality: 
- it introduces CyclicLRs (that I am fond of and convinced that they work great) 
- it ties everything together. 
- it adds functionality for the user to read all these from the model_config. 

BUT, we are missing a great oppty: to make things clean, testable and modular. 
See here for instance: 
- We do a choice to have a scheduler as a default while giving the user the oppty to decide if they want to :-)
- we keep everything in a big `if/elif` function..

We should extract at least: 
- one function to decide the scheduler 
```python
if 'lr_schedule' in model_config:
  lr_scheduler = get_scheduler(model_config)
```
- another to decide the optimizer
```
if 'optimizer' in model_config:
  optmizer = get_optimizer(model_config, lr_scheduler)
```
This way things will be clear and testable. And the functions easier to read imho!